<a href="https://colab.research.google.com/github/FaizaAli-DS/Python-for-Machine-Learning-with-Faiza/blob/all-about-AI-and-Data-Science/Lab_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 1: Python basics for AI
Generate Python code examples demonstrating fundamental programming concepts such as basic syntax, variables, data types, operators, input/output, control flow (conditionals and loops), functions, and common data structures (lists, tuples, dictionaries), along with a summary of these concepts.

## Implement: Introduction and Comments

### Subtask:
Generate Python code demonstrating basic syntax and the use of comments.


**Reasoning**:
Generate a Python code block that includes a multi-line comment, a single-line comment, and a print statement as requested in the subtask instructions.



In [1]:
'''
This script demonstrates fundamental Python concepts.
It shows how to use basic syntax and different types of comments.
'''

# This print statement displays a greeting message to the console.
print("Hello, Python!")

Hello, Python!


## Implement: Variables and Data Types

### Subtask:
Generate Python code to declare variables and assign values of different data types (int, float, str, bool).


**Reasoning**:
The subtask requires generating Python code to declare variables of different data types (int, float, str, bool) and then printing their values and types. This code block will fulfill all parts of the instructions.



In [2]:
# 1. Declare an integer variable
age = 30

# 2. Declare a float variable
price = 19.99

# 3. Declare a string variable
name = "Alice"

# 4. Declare a boolean variable
is_active = True

# 5. Print the value and data type of each variable
print(f"Variable: {age}, Type: {type(age)}")
print(f"Variable: {price}, Type: {type(price)}")
print(f"Variable: {name}, Type: {type(name)}")
print(f"Variable: {is_active}, Type: {type(is_active)}")

Variable: 30, Type: <class 'int'>
Variable: 19.99, Type: <class 'float'>
Variable: Alice, Type: <class 'str'>
Variable: True, Type: <class 'bool'>


## Implement: Operators

### Subtask:
Generate Python code demonstrating arithmetic, comparison, logical, and assignment operators.


**Reasoning**:
This code block will define two numeric variables and two boolean variables, then proceed to demonstrate arithmetic, comparison, logical, and assignment operators by applying them to these variables and printing the results, fulfilling all parts of the subtask instructions.

**Reasoning**:
The subtask requires generating Python code to demonstrate arithmetic, comparison, logical, and assignment operators. This code block will initialize the necessary variables and then systematically apply and print the results of each type of operator.



In [3]:
num1 = 10
num2 = 3
x = True
y = False

print("--- Arithmetic Operators ---")
print(f"Addition: {num1} + {num2} = {num1 + num2}")
print(f"Subtraction: {num1} - {num2} = {num1 - num2}")
print(f"Multiplication: {num1} * {num2} = {num1 * num2}")
print(f"Division: {num1} / {num2} = {num1 / num2}") # Float division
print(f"Modulus: {num1} % {num2} = {num1 % num2}")
print(f"Exponentiation: {num1} ** {num2} = {num1 ** num2}")
print(f"Floor Division: {num1} // {num2} = {num1 // num2}")

print("\n--- Comparison Operators ---")
print(f"Equal to: {num1} == {num2} is {num1 == num2}")
print(f"Not equal to: {num1} != {num2} is {num1 != num2}")
print(f"Greater than: {num1} > {num2} is {num1 > num2}")
print(f"Less than: {num1} < {num2} is {num1 < num2}")
print(f"Greater than or equal to: {num1} >= {num2} is {num1 >= num2}")
print(f"Less than or equal to: {num1} <= {num2} is {num1 <= num2}")

print("\n--- Logical Operators ---")
print(f"x and y: {x} and {y} is {x and y}")
print(f"x or y: {x} or {y} is {x or y}")
print(f"not x: not {x} is {not x}")

print("\n--- Assignment Operators ---")
a = 5
print(f"Initial value of a: {a}")
a += 3
print(f"a += 3: {a}")
a -= 2
print(f"a -= 2: {a}")
a *= 4
print(f"a *= 4: {a}")
a /= 2
print(f"a /= 2: {a}")
a %= 3
print(f"a %= 3: {a}")
a = 2 # Reset for exponentiation
a **= 3
print(f"a **= 3: {a}")
a = 10 # Reset for floor division
a //= 3
print(f"a //= 3: {a}")

--- Arithmetic Operators ---
Addition: 10 + 3 = 13
Subtraction: 10 - 3 = 7
Multiplication: 10 * 3 = 30
Division: 10 / 3 = 3.3333333333333335
Modulus: 10 % 3 = 1
Exponentiation: 10 ** 3 = 1000
Floor Division: 10 // 3 = 3

--- Comparison Operators ---
Equal to: 10 == 3 is False
Not equal to: 10 != 3 is True
Greater than: 10 > 3 is True
Less than: 10 < 3 is False
Greater than or equal to: 10 >= 3 is True
Less than or equal to: 10 <= 3 is False

--- Logical Operators ---
x and y: True and False is False
x or y: True or False is True
not x: not True is False

--- Assignment Operators ---
Initial value of a: 5
a += 3: 8
a -= 2: 6
a *= 4: 24
a /= 2: 12.0
a %= 3: 0.0
a **= 3: 8
a //= 3: 3


## Implement: Input and Output

### Subtask:
Generate Python code to get user input and print output to the console.


**Reasoning**:
The subtask requires generating Python code to get user input for name and age, convert the age to an integer, and then print a personalized greeting to the console. The code block will fulfill all these requirements.



In [4]:
import sys

# 1. Prompt the user to enter their name
user_name = input("Enter your name: ")

# 2. Prompt the user to enter their age and convert it to an integer
# Using a try-except block to handle potential ValueError if input is not a valid integer.
try:
    user_age = int(input("Enter your age: "))
except ValueError:
    print("Invalid age entered. Please enter a number.", file=sys.stderr)
    user_age = 0 # Assign a default or handle as appropriate

# 3. Print a personalized greeting message
if user_age != 0: # Only print greeting if age was valid or handled
    print(f"Hello, {user_name}! You are {user_age} years old.")
else:
    print(f"Hello, {user_name}! We couldn't determine your age.")

Enter your name: faiza
Enter your age: 30
Hello, faiza! You are 30 years old.


## Implement: Control Flow - Conditional Statements

### Subtask:
Generate Python code demonstrating 'if', 'elif', and 'else' statements.


**Reasoning**:
This code block addresses the subtask by demonstrating conditional statements (`if`, `elif`, `else`). It initializes a `score` variable and then uses a series of `if-elif-else` checks to determine and print a grade based on the score, fulfilling all the specified instructions.

**Reasoning**:
This code block will implement the conditional statements as instructed, initializing a score, and then using `if`, `elif`, and `else` to print appropriate messages based on the score's value.



In [5]:
# 1. Initialize a variable 'score'
score = 75 # You can change this value to test different cases

print(f"Checking score: {score}")

# 2. Use an if statement
if score >= 90:
    print("Excellent!")
# 3. Use an elif statement
elif score >= 70:
    print("Good job!")
# 4. Use another elif statement
elif score >= 50:
    print("Pass.")
# 5. Use an else statement
else:
    print("Fail.")

# Testing with different scores (optional, for demonstration)
print("\n--- Testing with different scores ---")

def check_score(s):
    print(f"Checking score: {s}")
    if s >= 90:
        print("Excellent!")
    elif s >= 70:
        print("Good job!")
    elif s >= 50:
        print("Pass.")
    else:
        print("Fail.")

check_score(95) # Should print Excellent!
check_score(80) # Should print Good job!
check_score(60) # Should print Pass.
check_score(40) # Should print Fail.

Checking score: 75
Good job!

--- Testing with different scores ---
Checking score: 95
Excellent!
Checking score: 80
Good job!
Checking score: 60
Pass.
Checking score: 40
Fail.


## Implement: Control Flow - Loops

### Subtask:
Generate Python code demonstrating 'for' and 'while' loops.

## Implement: Control Flow - Loops

### Subtask:
Generate Python code demonstrating 'for' and 'while' loops.

#### Instructions
1. Use a `for` loop to iterate over a list of items (e.g., numbers, strings) and print each item.
2. Use the `range()` function with a `for` loop to iterate a specific number of times.
3. Use a `while` loop to repeatedly execute a block of code as long as a condition is true. Include a counter and a breaking condition to avoid infinite loops.

#### Instructions
1. Use a `for` loop to iterate over a list of items (e.g., numbers, strings) and print each item.
2. Use the `range()` function with a `for` loop to iterate a specific number of times.
3. Use a `while` loop to repeatedly execute a block of code as long as a condition is true. Include a counter and a breaking condition to avoid infinite loops.

```markdown
**Reasoning**:
This code block will fulfill the subtask requirements by demonstrating both 'for' and 'while' loops. It will include a 'for' loop iterating over a list, another 'for' loop using `range()`, and a 'while' loop with a counter and a break condition to prevent an infinite loop, covering all points in the instructions.
```

**Reasoning**:
This code block will fulfill the subtask requirements by demonstrating both 'for' and 'while' loops. It will include a 'for' loop iterating over a list, another 'for' loop using `range()`, and a 'while' loop with a counter and a break condition to prevent an infinite loop, covering all points in the instructions.

**Reasoning**:
This code block will fulfill the subtask requirements by demonstrating both 'for' and 'while' loops. It will include a 'for' loop iterating over a list, another 'for' loop using `range()`, and a 'while' loop with a counter and a break condition to prevent an infinite loop, covering all points in the instructions.

**Reasoning**:
This code block will fulfill the subtask requirements by demonstrating both 'for' and 'while' loops. It will include a 'for' loop iterating over a list, another 'for' loop using `range()`, and a 'while' loop with a counter and a break condition to prevent an infinite loop, covering all points in the instructions.

**Reasoning**:
This code block will implement the 'for' and 'while' loops as instructed, demonstrating iteration over a list, iteration using `range()`, and a `while` loop with a controlled exit condition.



In [6]:
# 1. For loop iterating over a list of items
print("--- For Loop (List Iteration) ---")
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(f"I like {fruit}")

# 2. For loop using range() to iterate a specific number of times
print(
"\n--- For Loop (Range) ---")
for i in range(5): # Iterates from 0 to 4
    print(f"Count: {i}")

# 3. While loop with a counter and a breaking condition
print("\n--- While Loop ---")
count = 0
while count < 3:
    print(f"Current count: {count}")
    count += 1
print("While loop finished.")

--- For Loop (List Iteration) ---
I like apple
I like banana
I like cherry

--- For Loop (Range) ---
Count: 0
Count: 1
Count: 2
Count: 3
Count: 4

--- While Loop ---
Current count: 0
Current count: 1
Current count: 2
While loop finished.


## Implement: Functions

### Subtask:
Generate Python code to define and call a simple function with parameters and a return value.


#### Instructions
1. Define a function that takes two numeric parameters, calculates their sum, and returns the result.
2. Call the defined function with two different sets of arguments.
3. Print the results returned by the function calls, clearly indicating what each result represents.

**Reasoning**:
This code block will define a Python function named `add_numbers` that accepts two parameters, calculates their sum, and returns the result, thereby fulfilling instruction 1. Subsequently, it will call this function with two different sets of arguments and print the results with clear descriptions, addressing instructions 2 and 3.

**Reasoning**:
This code block will define a Python function named `add_numbers` that accepts two parameters, calculates their sum, and returns the result, thereby fulfilling instruction 1. Subsequently, it will call this function with two different sets of arguments and print the results with clear descriptions, addressing instructions 2 and 3.

**Reasoning**:
This code block will define a Python function named `add_numbers` that accepts two parameters, calculates their sum, and returns the result, thereby fulfilling instruction 1. Subsequently, it will call this function with two different sets of arguments and print the results with clear descriptions, addressing instructions 2 and 3.

**Reasoning**:
This code block will define a Python function named `add_numbers` that accepts two parameters, calculates their sum, and returns the result, thereby fulfilling instruction 1. Subsequently, it will call this function with two different sets of arguments and print the results with clear descriptions, addressing instructions 2 and 3.

**Reasoning**:
This code block will define a Python function named `add_numbers` that accepts two parameters, calculates their sum, and returns the result, thereby fulfilling instruction 1. Subsequently, it will call this function with two different sets of arguments and print the results with clear descriptions, addressing instructions 2 and 3.

**Reasoning**:
This code block will define a Python function named `add_numbers` that accepts two parameters, calculates their sum, and returns the result, thereby fulfilling instruction 1. Subsequently, it will call this function with two different sets of arguments and print the results with clear descriptions, addressing instructions 2 and 3.


**Reasoning**:
This code block will define a Python function named `add_numbers` that accepts two parameters, calculates their sum, and returns the result, thereby fulfilling instruction 1. Subsequently, it will call this function with two different sets of arguments and print the results with clear descriptions, addressing instructions 2 and 3.


**Reasoning**:
This code block will define a Python function named `add_numbers` that accepts two parameters, calculates their sum, and returns the result, thereby fulfilling instruction 1. Subsequently, it will call this function with two different sets of arguments and print the results with clear descriptions, addressing instructions 2 and 3.


**Reasoning**:
This code block will define a Python function as per the instructions, demonstrating how to define a function with parameters and a return value, and then call it with different arguments, finally printing the results.



In [7]:
# 1. Define a function that takes two numeric parameters, calculates their sum, and returns the result.
def add_numbers(num1, num2):
    """
    This function takes two numbers as input and returns their sum.
    """
    return num1 + num2

# 2. Call the defined function with two different sets of arguments.
result1 = add_numbers(10, 5)
result2 = add_numbers(3.5, 2.1)

# 3. Print the results returned by the function calls.
print(f"The sum of 10 and 5 is: {result1}")
print(f"The sum of 3.5 and 2.1 is: {result2}")

The sum of 10 and 5 is: 15
The sum of 3.5 and 2.1 is: 5.6


## Implement: Data Structures

### Subtask:
Generate Python code demonstrating common data structures: lists, tuples, and dictionaries.

## Implement: Data Structures

### Subtask:
Generate Python code demonstrating common data structures: lists, tuples, and dictionaries.

## Implement: Data Structures

### Subtask:
Generate Python code demonstrating common data structures: lists, tuples, and dictionaries.

## Implement: Data Structures

### Subtask:
Generate Python code demonstrating common data structures: lists, tuples, and dictionaries.

## Implement: Data Structures

### Subtask:
Generate Python code demonstrating common data structures: lists, tuples, and dictionaries.

## Implement: Data Structures

### Subtask:
Generate Python code demonstrating common data structures: lists, tuples, and dictionaries.

## Implement: Data Structures

### Subtask:
Generate Python code demonstrating common data structures: lists, tuples, and dictionaries.


#### Instructions
1. Initialize a list with at least three elements (e.g., numbers or strings).
2. Add a new element to the end of the list using an appropriate method.
3. Remove an element from the list.
4. Access and print an element from the list using its index.
5. Initialize a tuple with at least three elements.
6. Access and print an element from the tuple using its index.
7. Initialize a dictionary with at least three key-value pairs (e.g., student name and age).
8. Add a new key-value pair to the dictionary.
9. Update the value of an existing key in the dictionary.
10. Access and print a value from the dictionary using its key.
11. Remove a key-value pair from the dictionary.

#### Instructions
1. Initialize a list with at least three elements (e.g., numbers or strings).
2. Add a new element to the end of the list using an appropriate method.
3. Remove an element from the list.
4. Access and print an element from the list using its index.
5. Initialize a tuple with at least three elements.
6. Access and print an element from the tuple using its index.
7. Initialize a dictionary with at least three key-value pairs (e.g., student name and age).
8. Add a new key-value pair to the dictionary.
9. Update the value of an existing key in the dictionary.
10. Access and print a value from the dictionary using its key.
11. Remove a key-value pair from the dictionary.

#### Instructions
1. Initialize a list with at least three elements (e.g., numbers or strings).
2. Add a new element to the end of the list using an appropriate method.
3. Remove an element from the list.
4. Access and print an element from the list using its index.
5. Initialize a tuple with at least three elements.
6. Access and print an element from the tuple using its index.
7. Initialize a dictionary with at least three key-value pairs (e.g., student name and age).
8. Add a new key-value pair to the dictionary.
9. Update the value of an existing key in the dictionary.
10. Access and print a value from the dictionary using its key.
11. Remove a key-value pair from the dictionary.

#### Instructions
1. Initialize a list with at least three elements (e.g., numbers or strings).
2. Add a new element to the end of the list using an appropriate method.
3. Remove an element from the list.
4. Access and print an element from the list using its index.
5. Initialize a tuple with at least three elements.
6. Access and print an element from the tuple using its index.
7. Initialize a dictionary with at least three key-value pairs (e.g., student name and age).
8. Add a new key-value pair to the dictionary.
9. Update the value of an existing key in the dictionary.
10. Access and print a value from the dictionary using its key.
11. Remove a key-value pair from the dictionary.

**Reasoning**:
The previous steps defined the instructions for demonstrating common data structures. Now, I will generate the Python code to implement these instructions for lists, tuples, and dictionaries, covering initialization, modification, and access as required by the subtask.



In [8]:
print("--- Lists ---")
# 1. Initialize a list with at least three elements
my_list = [10, 20, 30]
print(f"Initial list: {my_list}")

# 2. Add a new element to the end of the list
my_list.append(40)
print(f"List after appending 40: {my_list}")

# 3. Remove an element from the list
my_list.remove(20) # Removes the first occurrence of 20
print(f"List after removing 20: {my_list}")

# 4. Access and print an element from the list using its index
print(f"Element at index 1: {my_list[1]}") # Accesses 30


print("\n--- Tuples ---")
# 5. Initialize a tuple with at least three elements
my_tuple = ("apple", "banana", "cherry")
print(f"Initial tuple: {my_tuple}")

# 6. Access and print an element from the tuple using its index
print(f"Element at index 0: {my_tuple[0]}") # Accesses "apple"
# Note: Tuples are immutable, so elements cannot be added, removed, or updated after creation.


print("\n--- Dictionaries ---")
# 7. Initialize a dictionary with at least three key-value pairs
my_dict = {"name": "Alice", "age": 25, "city": "New York"}
print(f"Initial dictionary: {my_dict}")

# 8. Add a new key-value pair to the dictionary
my_dict["occupation"] = "Engineer"
print(f"Dictionary after adding 'occupation': {my_dict}")

# 9. Update the value of an existing key in the dictionary
my_dict["age"] = 26
print(f"Dictionary after updating 'age': {my_dict}")

# 10. Access and print a value from the dictionary using its key
print(f"Alice's city: {my_dict['city']}")

# 11. Remove a key-value pair from the dictionary
del my_dict["city"]
print(f"Dictionary after removing 'city': {my_dict}")

--- Lists ---
Initial list: [10, 20, 30]
List after appending 40: [10, 20, 30, 40]
List after removing 20: [10, 30, 40]
Element at index 1: 30

--- Tuples ---
Initial tuple: ('apple', 'banana', 'cherry')
Element at index 0: apple

--- Dictionaries ---
Initial dictionary: {'name': 'Alice', 'age': 25, 'city': 'New York'}
Dictionary after adding 'occupation': {'name': 'Alice', 'age': 25, 'city': 'New York', 'occupation': 'Engineer'}
Dictionary after updating 'age': {'name': 'Alice', 'age': 26, 'city': 'New York', 'occupation': 'Engineer'}
Alice's city: New York
Dictionary after removing 'city': {'name': 'Alice', 'age': 26, 'occupation': 'Engineer'}


## Final Task

### Subtask:
Summarize the basic Python concepts covered and provide any final tips for learning Python.


## Summary:

### Data Analysis Key Findings

*   **Basic Syntax and Comments**: Python's fundamental syntax was demonstrated through a simple print statement. Both single-line (`#`) and multi-line (`'''...'''`) comments were correctly implemented without affecting code execution.
*   **Variables and Data Types**: Variables were successfully declared and assigned values of various types: integer (e.g., `age = 30`), float (e.g., `price = 19.99`), string (e.g., `name = "Alice"`), and boolean (e.g., `is_active = True`). The `type()` function confirmed Python's dynamic typing.
*   **Operators**:
    *   **Arithmetic**: All standard operations (addition, subtraction, multiplication, float division, modulus, exponentiation, floor division) were shown, with examples like `10 / 3` yielding `3.3333333333333335` and `10 // 3` yielding `3`.
    *   **Comparison**: Operators such as `==`, `!=`, `>`, `<`, `>=`, `<=` correctly compared values, producing boolean results (e.g., `10 == 3` is `False`).
    *   **Logical**: `and`, `or`, and `not` operators were demonstrated with boolean variables, yielding expected outcomes (e.g., `True and False` is `False`).
    *   **Assignment**: Compound assignment operators like `+=`, `-=`, `*=`, `/=`, `%=`, `**=`, `//=` were used to modify variable values efficiently (e.g., `a += 3` changed `a` from $5$ to $8$).
*   **Input and Output**: User input was obtained using `input()`, and output was displayed using `print()` with f-strings for formatted messages (e.g., "Hello, faiza! You are 30 years old."). Error handling for invalid input (e.g., `ValueError` for age conversion) was also included.
*   **Control Flow (Conditionals)**: `if`, `elif`, and `else` statements were used to implement conditional logic, assigning grades based on a `score`. A score of $75$ correctly resulted in "Good job!", while $95$ yielded "Excellent!", $60$ "Pass.", and $40$ "Fail.".
*   **Control Flow (Loops)**:
    *   **For Loops**: Iteration over a list (e.g., printing "I like apple") and iteration a specific number of times using `range()` (e.g., "Count: $0$ " to "Count: $4$ ") were demonstrated.
    *   **While Loops**: A `while` loop executed based on a condition, with a counter and a break condition to prevent infinite loops (e.g., printing "Current count: $0$" to "Current count: $2$ ").
*   **Functions**: A custom function `add_numbers()` was defined to accept two parameters, calculate their sum, and return the result. Calling it with `add_numbers(10, 5)` returned $15$, and `add_numbers(3.5, 2.1)` returned $5.6$.
*   **Data Structures**:
    *   **Lists**: Operations demonstrated included initialization (`[10, 20, 30]`), appending (`.append(40)` resulting in `[10, 20, 30, 40]`), removing (`.remove(20)` resulting in `[10, 30, 40]`), and accessing elements by index (e.g., `my_list[1]` yielded $30$).
    *   **Tuples**: Initialization (`("apple", "banana", "cherry")`) and accessing elements by index (`my_tuple[0]` yielded "apple") were shown, highlighting their immutability.
    *   **Dictionaries**: Operations included initialization (`{"name": "Alice", "age": 25}`), adding a new key-value pair (`"occupation": "Engineer"`), updating an existing value (`"age"` from $25$ to $26$), accessing values by key (e.g., `my_dict['city']` yielded "New York"), and removing a key-value pair (`del my_dict["city"]`).

### Insights or Next Steps

*   Python's intuitive syntax, combined with its rich set of built-in data types and control structures, makes it an accessible language for beginners to grasp fundamental programming concepts.
*   To further enhance understanding, learners should practice implementing these concepts in small projects, exploring more advanced features within each topic (e.g., nested loops, function arguments, list comprehensions, dictionary methods) and debugging common errors.
